File: MicroTDCore/Sources/MicroTDCore/Definitions/RelicDatabase.swift
// RelicDatabase.swift
// Efficient lookup for RelicDefinitions

import Foundation

public struct RelicDatabase: Sendable {
    public let all: [RelicDef]
    public let byID: [String: RelicDef]
    
    public init(relicDefs: [RelicDef]) {
        self.all = relicDefs
        self.byID = Dictionary(uniqueKeysWithValues: relicDefs.map { ($0.id, $0) })
    }
    
    // Convenience for GameDefinitions
    public init(definitions: RelicDefinitions) {
        self.init(relicDefs: definitions.all)
    }
    
    public func relic(id: String) -> RelicDef? {
        return byID[id]
    }
}

File: expo/src/core/definitions/RelicDatabase.ts
// RelicDatabase.ts
// Efficient relic lookup — ported from MicroTDCore/Definitions/RelicDatabase.swift

import type { RelicDef } from './RelicDef';
import { RelicDefinitions } from './RelicDef';

/**
 * Indexed relic database for O(1) ID lookups
 */
export class RelicDatabase {
    public readonly all: RelicDef[];
    private readonly byID: Map<string, RelicDef>;

    constructor(definitions: RelicDefinitions) {
        this.all = definitions.relics;
        this.byID = new Map();
        for (const relic of definitions.relics) {
            this.byID.set(relic.id, relic);
        }
    }

    /**
     * Find relic by ID (O(1) lookup)
     */
    relic(id: string): RelicDef | undefined {
        return this.byID.get(id);
    }
}
